AzureCore/Source/Pipeline/Policies/LoggingPolicy.swift

```diff
@@ -0,0 +1,133 @@
+//
+//  LoggingPolicy.swift
+//  AzureCore
+//
+//  Created by Brandon Siegel on 10/2/19.
+//  Copyright © 2019 Azure SDK Team. All rights reserved.
+//
+
+import Foundation
+
+public class DefaultLoggingPolicy: PipelineStageProtocol {
+
+    public var next: PipelineStageProtocol?
+    private let log: PipelineLogger
+    private lazy var attachmentRegex = NSRegularExpression("attachment; ?filename=([\"\\w.]+)",
+                                                           options: .caseInsensitive)
+
+    public init(logger: PipelineLogger? = nil) {
+        if let logger = logger {
+            self.log = logger
+        } else if #available(macOS 10.12, iOS 10.0, watchOS 3.0, tvOS 10.0, *) {
+            self.log = PipelineOSLogAdapter()
+        } else {
+            self.log = PipelineNSLogger()
+        }
+    }
+
+    public func onRequest(_ request: inout PipelineRequest) {
+        let req = request.httpRequest
+        log.debug("Request URL: \(req.url)")
+        log.debug("Request method: \(req.httpMethod.rawValue)")
+        log.debug("Request headers:")
+        for (header, value) in req.headers {
+            log.debug("    \(header): \(value)")
+        }
+        log.debug("Request body:")
+        log.debug(loggableBodyFor(headers: req.headers, bodyText: req.text))
+    }
+
+    public func onResponse(_ response: inout PipelineResponse) {
+        guard let res = response.httpResponse else {
+            log.debug("Failed to log response")
+            return
+        }
+
+        if let statusCode = res.statusCode {
+            log.debug("Response status: \(statusCode)")
+        }
+        log.debug("Response headers:")
+        for (header, value) in res.headers {
+            log.debug("    \(header): \(value)")
+        }
+        log.debug("Response content:")
+        log.debug(loggableBodyFor(headers: res.headers, bodyText: res.text))
+    }
+
+    public func onError(_ error: Error) -> Bool {
+        log.error(error.localizedDescription)
```

Comment:
This should be fine for transport errors. We should consider when we would want a service error to be detected and whether a deserialized error should be passed through the onError section of the logging policy. The way current pipelines are set up, I don't believe they would.

Reply:
@bsiegel thoughts on this? Also happy to merge this now and defer this discussion.

Reply:
I think it's outside the scope of logging, but yes we should have a chat about how to represent protocol errors vs. transport errors (and further, retryable errors vs. fatal errors). I'd want to understand how the other languages currently make this distinction so I should probably take some time to read thru their codebases / guidelines on this topic.